contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+### Staged Roll Outs
+Users can create a canary using `kubectl apply`. The only difference between a
+ [canary](#canaries) and a staged roll out is that the `.Spec.GenerationPartition` 
+ is set to value less than `.Spec.Replicas - 1`.
+ 
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+      generationParition: 2
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Staged roll outs can be used to roll out a configuration, image, or resource 
+update to some portion of the fleet maintained by the StatefulSet prior to 
+updating the entire fleet. It is useful to support linear, geometric, and 
+exponential roll out of an update. Users can modify the 
+`.Spec.GenerationPartition` to allow the roll out to progress.
+
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+      generationParition: 1
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+### Rollbacks
+To rollback an update, users can use the `kubectl rollout` command.
+
+The command below will roll back the `web` StatefulSet to the previous revision in 
+its history. If a roll out is in progress, it will stop deploying the target 
+revision, and roll back to the current revision.
+
+```shell
+kubectl rollout undo statefulset web
+```
+
+### Rolling Forward
+Rolling back is usually the safest, and often the fastest, strategy to mitigate
+deployment failure, but rolling forward is sometimes the only practical solution 
+for stateful applications (e.g. A users has a minor configuration error but has 
+already modified the storage format for the application). Users can use 
+sequential `kubectl apply`'s to update the `.Status.TargetGeneration` of a 
```

Comment:
Users shouldn't be able to mutate the status. A user should simply update the spec with the correct configuration, the api server increments spec.templateGeneration, the controller sees that spec.templateGeneration != status.targetTemplateGeneration so it updates status.targetTemplateGeneration and creates a new PodTemplate to rollover the previous Pods.